pkg/reconciler/apis/apiexportendpointslice/apiexportendpointslice_controller.go

```diff
@@ -0,0 +1,247 @@
+
+// NewController returns a new controller for APIExportEndpointSlices.
+// ClusterWorkspaceShards and APIExports are read from the cache server
+func NewController(
+	apiExportEndpointSliceInformer apisinformers.APIExportEndpointSliceClusterInformer,
+	clusterWorkspaceShardInformer tenancyinformers.ClusterWorkspaceShardClusterInformer,
+	apiExportClusterInformer apisinformers.APIExportClusterInformer,
+	kcpClusterClient kcpclientset.ClusterInterface,
+) (*controller, error) {
+	queue := workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), ControllerName)
+
+	c := &controller{
+		queue:                         queue,
+		apiExportEndpointSliceLister:  apiExportEndpointSliceInformer.Lister(),
+		apiExportEndpointSliceIndexer: apiExportEndpointSliceInformer.Informer().GetIndexer(),
+		kcpClusterClient:              kcpClusterClient,
+		listClusterWorkspaceShards: func() ([]*tenancyv1alpha1.ClusterWorkspaceShard, error) {
+			return clusterWorkspaceShardInformer.Lister().List(labels.Everything())
+		},
+		getAPIExport: func(clusterName logicalcluster.Name, name string) (*apisv1alpha1.APIExport, error) {
+			return apiExportClusterInformer.Lister().Cluster(clusterName).Get(name)
+		},
+		commit: committer.NewCommitter[*APIExportEndpointSlice, Patcher, *APIExportEndpointSliceSpec, *APIExportEndpointSliceStatus](kcpClusterClient.ApisV1alpha1().APIExportEndpointSlices()),
+	}
+
+	apiExportEndpointSliceInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			c.enqueueAPIExportEndpointSlice(obj)
+		},
+		UpdateFunc: func(_, newObj interface{}) {
+			c.enqueueAPIExportEndpointSlice(newObj)
+		},
+		DeleteFunc: func(obj interface{}) {
+			c.enqueueAPIExportEndpointSlice(obj)
+		},
+	})
+
+	clusterWorkspaceShardInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				c.enqueueAllAPIExportEndpointSlices(obj)
+			},
+			UpdateFunc: func(_, newObj interface{}) {
+				c.enqueueAllAPIExportEndpointSlices(newObj)
+			},
+			DeleteFunc: func(obj interface{}) {
+				c.enqueueAllAPIExportEndpointSlices(obj)
+			},
+		},
+	)
+
+	return c, nil
+}
+
+type APIExportEndpointSlice = apisv1alpha1.APIExportEndpointSlice
+type APIExportEndpointSliceSpec = apisv1alpha1.APIExportEndpointSliceSpec
+type APIExportEndpointSliceStatus = apisv1alpha1.APIExportEndpointSliceStatus
+type Patcher = apisv1alpha1client.APIExportEndpointSliceInterface
+type Resource = committer.Resource[*APIExportEndpointSliceSpec, *APIExportEndpointSliceStatus]
+type CommitFunc = func(context.Context, *Resource, *Resource) error
+
+// controller reconciles APIExportEndpointSlices. It ensures that the shard endpoints are populated
+//  in the status of every APIExportEndpointSlices.
+type controller struct {
+	queue workqueue.RateLimitingInterface
+
+	apiExportEndpointSliceLister  apislisters.APIExportEndpointSliceClusterLister
+	apiExportEndpointSliceIndexer cache.Indexer
```

Comment:
Please store function pointers instead of listers/indexers, e.g. `listAPIExportEndpointSlices func(...)`

Reply:
amended. Mind however that listers and Indexers are stored in both apiexport and apibinding controllers.

Reply:
Thanks. Those controllers were written before we started using the function pointer approach. We'll come back and update them.